File: backend/model/challenge.go
package model

import (
	"github.com/satori/go.uuid"
)

// Challenge describes a single challenge
type Challenge struct {
	ID      uuid.UUID `json:"id"`
	Text    string    `json:"text"`
	Choices []*User   `json:"choices"`
	Author  *User     `json:"author"`
}

File: frontend/src/state/Challenge.js
import { types, getRoot } from "mobx-state-tree";
import Author from "state/Author";
import { Views } from "Constants";

const Challenge = types
  .model({
    id: types.string,
    text: types.string,
    choices: types.array(Author),
    author: Author,
    answeredAuthorId: types.maybeNull(types.string)
  })
  .actions(self => ({
    answer(authorId) {
      const { ChallengeStore, ViewStore } = getRoot(self);

      self.answeredAuthorId = authorId;

      if (!ChallengeStore.currentChallenge) {
        ViewStore.presentView(Views.SCORE);
      }
    }
  }))
  .views(self => ({
    get isAnswered() {
      return self.answeredAuthorId !== null;
    },

    get isCorrect() {
      return self.answeredAuthorId === self.author.id;
    }
  }));

export default Challenge;
